### ISSUE TITLE
Alternatives for current ensure_communicating

### ISSUE THREAD

**Issue by fjetter** [dask/distributed#6497]
We've recently been refactoring the worker state machine (https://github.com/dask/distributed/issues/5736)

The new system works great for single-task-based transitions but feels awkward as soon as we're trying to treat groups or batches of tasks similarly which is particularly important for fetching remote data. We are trying to not perform a network request per key but rather try to perform a request per worker and try to get as many keys of it as reasonable.

This mechanism worked so far by running `Worker.(_)ensure_communicating` frequently which iterates over all tasks and schedules a fetch/gather_dep (pseudo code below)

```python
def ensure_communicating(...):
    skip = []
    while self.data_needed and free_slots_for_fetch:
        ts = self.data_needed.pop() # highest priority task
        worker = pick_good_worker(ts) # not busy, maybe localhost, etc.
        if not worker:
            skip.add(ts)
        more_keys_on_worker = select_keys_for_gather(worker)
        schedule_gather_dep(ts, more_keys_on_worker, worker) # This schedules a coroutine and actually fetches the data
    self.data_needed.update(skip) # data needed is a heap so this is non-linear
```

This has a couple of problems
1. It's unclear when to run this function. Past iterations just ran this frequently (on every incoming request and in various code places where we guess there may be something that needs transitioning
2. We're iterating over all keys, all the time. This is bad for performance although not critical on the worker https://github.com/dask/distributed/pull/6462#issuecomment-1138764541
3. If we're calling this too soon, e.g. within a released->fetch transition, we may get poor scheduling decisions by fetching tasks prematurely since we didn't work off all fetch recommendations before calling this. For this reason, the `EnsureCommunicatingAfterTransitions ` instruction was introduced (see also https://github.com/dask/distributed/pull/6462) to mimic earlier behavior

## Proposal

I think we can get rid of `Worker.data_needed` entirely. Since we're storing everything in heaps, by now, `Worker.data_needed_per_worker` carries everything we need if we change how ensure_communicating works which might have other interesting side effects.

To illustrate this, consider the following two helper functions

```python
def get_most_important_worker_to_fetch_from(self):
    """
    Get the remote worker that holds the _most important_ data we should fetch. This is typically inferred by looking at task priorities.
    This could also be used to modify what we consider most important worker to fetch from, e.g. due to load/business, memory pressure, etc.
    """

    # Linear in num_workers
    res = None
    prio = None
    for w, heap in self.data_needed_per_worker.items():
        nprio = heap.peek().priority
        if w not in self.busy_workers and (res is None or nprio > prio):
            res = w
            prio = nprio
    return res

def get_tasks_for_worker(self, w, limit=42):
    """Get a list of tasks on worker `w` such that combined they have at most `limit` estimated bytes"""
    # This is more or less the current `Worker._select_keys_for_gather`
    # Linear in number of tasks in local state fetch on remote worker
    res = []
    size = 0
    for ts in self.data_needed_per_worker[w]:
        size += ts.get_nbytes()
        res.append(ts)
        if size > limit:
            break
    return res
```

With these two helpers we should be able to construct a new, simplified version if we switch our loop from tasks to workers

```python
def _new_ensure_communicate_with_tasks_by_worker():
    if network_limits_not_reached and worker := self.get_most_important_worker_to_fetch_from():
        tasks_to_fetch = self.get_tasks_for_worker(worker)
        instruction = GatherDep(
            worker,
            tasks_to_fetch,
        )
```

or even change the `GatherDep` instruction to worker-only and delay the decision of what tasks to fetch until this instruction is acted on, which typically only happens after all transitions in `Worker.transitions` are worked off

```python
def _new_ensure_communicate_with_tasks_by_worker():
    if network_limits_not_reached and worker := self.get_most_important_worker_to_fetch_from():
        instruction = GatherDep(worker)

def gather_dep(self, worker):
    tasks_to_fetch = self.get_tasks_for_worker(worker)
    ...
```

cc @crusaderky

**Comment by crusaderky** [comment]
While I'm warm to deduplicate data_needed vs. data_needed_per_worker, I don't see how any of this removes the problem about when to trigger `_new_ensure_communicate_with_tasks_by_worker`?

My suggestion:

Remove `data_needed_per_worker` entirely, and remain only with `data_needed`.
`in_flight_workers` becomes `{worker: GatherDep}` where the GatherDep instruction contains all tasks to be fetched from that worker.
Every iteration of `_ensure_communicating` fetches the topmost task from data_needed and adds it to the matching GatherDep for that worker. Then, it adds the GatherDep to instructions (possibly the same object multiple times).
`_ensure_communicating` stops iterating when any one of the following happens:

1. there are no tasks left in data_needed
2. the total nbytes in flight across all workers exceeds a global threshold
`self.comm_nbytes >= self.comm_threshold_bytes`
3. the number of pending GatherDep commands exceeds a threshold
`len(self.in_flight_workers) >= self.total_out_connections`

Individual GatherDep instructions stop being enriched when their individual nbytes exceed an individual threshold
`self.in_flight_workers[worker].nbytes >= self.target_message_size`; when that happens, tasks are moved to `skipped_worker_in_flight_or_busy` (read below).

At the top of `_handle_instructions`, there's a simple object deduplication:
```python
instructions = list({id(instr): instr for instr in instructions}.values())
```

`_ensure_communicating` is still called as soon as possible, e.g. at every transition to fetch. However, as a performance tweak, 
`skipped_worker_in_flight_or_busy` becomes an instance variable - a set which must lose keys every time one would otherwise remove keys from `data_needed`. Instead of being flushed back into `data_needed` at the end of each iteration of `_ensure_communicating`, it is partially flushed
1. every time a worker exits flight or busy state
2. every time update_who_has adds a new worker to a task in fetch state

flushing `skipped_worker_in_flight_or_busy` back into `data_needed` is O(n) to `len(skipped_worker_in_flight_or_busy)` and O(n*logn) to the number of tasks being added back in, but the key is that it happens a lot less frequently than now.

This closes https://github.com/dask/distributed/pull/6462#issuecomment-1138764541 for both the case of multiple fetches from the same event as well as for the case of multiple fetches from multiple commands in short succession, and will cause calling `_ensure_communicating` twice in a row to be computationally negligible.

**Comment by crusaderky** [comment]
The big difference between my proposal above and both the current system and the proposal to remove `data_needed` in the op is that my proposal respects global priority. This means that you'll end up with more, individually smaller, parallel tranfers from multiple workers as you'll be more likely to hit `self.comm_threshold_bytes` than `self.target_message_size`.

e.g. 
each key is nbytes=1
there are 99999 keys to fetch, uniformly distributed across 3 workers a, b, and c, with the same priority, and iteratively inserted in `data_needed` (a b c a b c a b c...)
self.comm_threshold_bytes=90
self.target_message_size=50

current system will fetch 50 keys from worker a and 40 keys from worker b
my proposal would fetch 30 keys from worker a, 30 from worker b, and 30 from worker c

**Comment by fjetter** [comment]
> While I'm warm to deduplicate data_needed vs. data_needed_per_worker, I don't see how any of this removes the problem about when to trigger _new_ensure_communicate_with_tasks_by_worker?

The proposal that includes an instruction `GatherDep(worker: str)` w/out explicit tasks would delay the collection of tasks until the instruction is executed.
Assuming that we perform all transitions before we execute any instructions, this would resolve our problem, wouldn't it?

**Comment by fjetter** [comment]
`ensure_communicating` would effectively be reduced to simply inferring whether or not we need to communicate and not _what_ to communicate.

**Comment by crusaderky** [comment]
> The proposal that includes an instruction GatherDep(worker: str) w/out explicit tasks would delay the collection of tasks until the instruction is executed.
> ensure_communicating would effectively be reduced to simply inferring whether or not we need to communicate and not what to communicate.

I see this as a major antipattern to the current worker state machine refactoring, as it would the whole contents of `_select_keys_for_gather` out of `WorkerState` and into `Worker.gather_dep`.
The check for `self.comm_nbytes < self.comm_threshold_bytes` would also need to move there too.

**Comment by fjetter** [comment]
> I see this as a major antipattern to the current worker state machine refactoring, as it would require the whole contents of _select_keys_for_gather to be moved out of WorkerState and into Worker.gather_dep. This in turn would require a new hack StartedGatheringEvent to allow for the transition from fetch to flight.
The check for self.comm_nbytes < self.comm_threshold_bytes would also need to move there too.

Good point.

---

Taking a step back, I'm wondering how we'd approach this from a TDD perspective. 

1. It should only involve `WorkerState` and not `Worker(Server)`
2. The event log should have roughly the following two items (modulo unrelevant kwargs)

This could be roughly written up as something like the following

```python
assert ComputeTaskEvent(key='f1', who_has={"f2": {"workerB"}, "f3", {"workerB"}, ...) in state.event_log
...
assert GatherDep(worker="workerB", to_gather={"f2", "f3"}) in state.instructions_log
```

and also

```python
assert AcquireReplicasEvent(who_has={"f2": {"workerB"}, "f3", {"workerB"}, ...) in state.event_log

assert GatherDep(worker="workerB", to_gather={"f2", "f3"}) in state.instructions_log

```

correct?

I'm wondering if we are already able to write such a test down, after https://github.com/dask/distributed/pull/6566


---

About the actual problem, I'm wondering if the order of transitioning cannot be used to influence this. I think right now the above requests would play out like (depth first)

- f1(released)
- f1(fetch)
- f1(flight)
- f2(released)
- f2(fetch)
- f2(flight)

Would it help if we instead performed a breadth-first transitioning?

- f1(released)
- f2(released)
- f1(fetch)
- f2(fetch)
- f1(flight)
- f2(flight)

**Comment by fjetter** [comment]
> could you comment on it?

sure, I'll need to give it another pass

> For testing, I envision something like

I would actually prefer to leave the task states out of this for this instance. I think the events and instructions should be enough and would make for great high level tests.

**Comment by crusaderky** [comment]
I've implemented 3B in #6462, and removed data_needed in #6587.

**Comment by crusaderky** [comment]
This issue is closed by:
- #6462 
- #6587 
- (no PR yet) ASV tests
- (no PR yet) further iterate on #6587, speeding up `_ensure_communicating` when there's nothing to do from O(n), where n is the number of workers holding replicas of tasks in fetch state, to O(1)
